indigo_app/forms.py

```diff
@@ -856,8 +856,94 @@ def save_changes(self):
+    update_import_tasks = forms.ChoiceField(choices=[('', ''), ('block', 'Block'), ('cancel', 'Cancel')], required=False)
+    import_task_description = forms.CharField(required=False)
+    gazette_task_works = forms.ModelMultipleChoiceField(queryset=Work.objects, required=False)
+    update_gazette_tasks = forms.ChoiceField(choices=[('', ''), ('block', 'Block'), ('cancel', 'Cancel')], required=False)
+    gazette_task_description = forms.CharField(required=False)
+    amendment_task_works = forms.ModelMultipleChoiceField(queryset=Work.objects, required=False)
+    amendment_task_amendments = forms.ModelMultipleChoiceField(queryset=Amendment.objects, required=False)
+    update_amendment_tasks = forms.ChoiceField(choices=[('', ''), ('block', 'Block'), ('cancel', 'Cancel')], required=False)
+    # TODO: add multichoice label dropdown per task type too
     approve = forms.BooleanField(required=False)
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self.is_valid():
+            self.add_amendment_task_description_fields(self.cleaned_data.get('works_in_progress', []))
+            self.full_clean()
+
+    def add_amendment_task_description_fields(self, works_in_progress):
+        for work in works_in_progress:
+            for amendment in work.amendments.all():
+                self.fields[f'amendment_task_description_{amendment.pk}'] = forms.CharField()
+
+    def save_changes(self, request):
+        for work in self.cleaned_data["works_in_progress"]:
+            work.approve(request.user, request)
+
+        # import tasks
+        if self.cleaned_data.get('import_task_works'):
+            # TODO: add the appropriate timeline date for Import tasks too?
+            import_tasks = [self.get_or_create_task(work=work, task_type='import-content', description=self.cleaned_data['import_task_description'], user=request.user) for work in self.cleaned_data['import_task_works']]
+            if self.cleaned_data.get('update_import_tasks'):
+                self.block_or_cancel_tasks(import_tasks, self.cleaned_data['update_import_tasks'], request.user)
+
+        # gazette tasks
+        if self.cleaned_data.get('gazette_task_works'):
+            gazette_tasks = [self.get_or_create_task(work=work, task_type='link-gazette', description=self.cleaned_data['gazette_task_description'], user=request.user) for work in self.cleaned_data['gazette_task_works']]
+            if self.cleaned_data.get('update_gazette_tasks'):
+                self.block_or_cancel_tasks(gazette_tasks, self.cleaned_data['update_gazette_tasks'], request.user)
+
+        # amendment tasks
+        if self.cleaned_data.get('amendment_task_works'):
+            amendment_tasks = []
+            for work in self.cleaned_data['amendment_task_works']:
+                for amendment in work.amendments.all():
+                    amendment_tasks.append(self.get_or_create_task(
```

Comment:
what is `timeline_date` doing? How different are these tasks from the tasks created during bulk import? Should we be re-using code, or is it good to start from scratch?

Reply:
`timeline_date` helps us target the appropriate task — in the case of amendments, it'll be the date of the amendment. We need this so that we don't end up creating a new Amendment task unnecessarily, nor update an existing one if it's tied to a different amendment.
There's potential to do much more with the date of course, but that's what it's doing here.

These tasks aren't wildly different from the bulk creator, but there are way fewer variables because we're dealing with live information, and way fewer task types. The way the tasks get updated now follows a different path too, and in the bulk creator it's all baked in together. I understood that the bulk creator will fall away, so I didn't really think about ways to reuse code between the two, but I'm happy to look into it. I do think it's simplest to do them separately though.